include/runtime/collections/rangemap.h

```diff
@@ -0,0 +1,380 @@
+    T start = r.first;
+    T end = r.second;
+    T rstart = t.root().first;
+    T rend = t.root().second;
+    if (rend <= start) {
+      // The candidate range is to the left of our target range, and may
+      // share a boundary. It is not relevant. Continue looking for
+      // relevant ranges to the right of the candidate range.
+      getDeletionRelevantRanges(t.right(), r, v);
+    } else if (end <= rstart) {
+      // The candidate range is to the right of our target range, and
+      // may share a boundary. It is not relevant. Continue looking for
+      // relevant ranges to the left of the candidate range.
+      getDeletionRelevantRanges(t.left(), r, v);
+    } else if (start < rstart && rend < end) {
+      // Our target range fully contains the candidate range, i.e. is
+      // larger. The candidate range is relevant, and there may be
+      // relevant ranges in both left and right directions.
+      getDeletionRelevantRanges(t.left(), r, v);
+      v.push_back(std::make_pair(t.root(), t.rootVal()));
+      getDeletionRelevantRanges(t.right(), r, v);
+    } else if (rstart <= start && end <= rend)
+    // The candidate range contains the target range, i.e. is equal or
+    // larger. It is the only relevant range, no need to look for more.
+    {
+      v.push_back(std::make_pair(t.root(), t.rootVal()));
+    } else if (rstart <= start) {
+      // The candidate range overlapps with the target range on the left
+      // side, but does not fully contain it. It is relevant, and we need
+      // to continue looking for relevant ranges to the right of the
+      // candidate range.
+      v.push_back(std::make_pair(t.root(), t.rootVal()));
+      getDeletionRelevantRanges(t.right(), r, v);
+    } else if (end <= rend) {
+      // The candidate range overlapps with the target range on the right
+      // side, but does not fully contain it. It is relevant, and we need
+      // to continue looking for relevant ranges to the left of the
+      // candidate range.
+      getDeletionRelevantRanges(t.left(), r, v);
+      v.push_back(std::make_pair(t.root(), t.rootVal()));
+    }
+  }
+
+public:
+  RangeMap()
+      : _map(RBTree<Range, V>()) { }
+
+  RangeMap(RBTree<Range, V> t)
+      : _map(t) { }
+
+  template <class I>
+  RangeMap(I b, I e) {
+    RangeMap m = RangeMap();
+    for_each(b, e, [&m](std::pair<Range, V> const &p) {
+      m = m.inserted(p.first, p.second);
+    });
+    _map = m._map;
+  }
+
+  RBTree<Range, V> getTreeMap() const { return _map; }
+
+  // Return the number of key ranges in the map.
+  size_t size() const { return _map.size(); }
+
+  // Return true if a range in the map contains the key.
+  bool containsKey(T k) const {
+    std::optional<std::pair<Range, V>> opt = getKeyValue(k);
+    if (opt.has_value())
+      return true;
+    return false;
+  }
+
+  // If the key is contained in any range in the map, return the value
+  // associated with the key.
+  std::optional<V> getValue(T k) const {
```

Comment:
See the upstream method comment - do we definitely need to copy the key to get the associated value?

Reply:
No. The key is now passed as reference, as discussed above.